commons/src/scanner/symbol/iterator/mod.rs

```diff
@@ -304,14 +349,54 @@ impl<'input> PeekingNext for SymbolIterator<'input> {
         Self: Sized,
         F: FnOnce(&Self::Item) -> bool,
     {
-        // Note: Not possible to restrict peek to return only symbols `next()` would return,
-        // because `peeking_next()` is needed in End- and PrefixMatcher.
-        // Using the same logic as in `next()` would result in endless loop inside `peeking_next()` => StackOverflow
+        if self.prefix_mismatch || self.end_reached() {
+            return None;
+        }
+
+        if !self.next_matching && !self.peek_matching {
+            if let Some(end_fn) = self.end_match.clone() {
+                let peek_index = self.peek_index();
+                self.peek_matching = true;
 
-        match &mut self.kind {
+                let end_matched = (end_fn)(self);
+
+                self.peek_matching = false;
+                self.set_peek_index(peek_index);
+
+                if end_matched {
+                    return None;
+                }
+            }
+        }
+
+        let peeked_symbol_opt = match &mut self.kind {
```

Comment:
I'm also questioning our decision to call this `kind`. Maybe `base`, `parent` or `wrapped`? In either case, there is some underlying iterator inside, not just kind 🤔.

But I don't think the name choice is too important here.

Reply:
`parent` sounds good